File: src/tui/countdown.py
from textual.app import App, ComposeResult
from textual.containers import Container
from textual.widgets import Digits, Footer, Header, Static
from textual.reactive import reactive
from core.formatting import format_time
from core.timer import Countdown
from tui.theme import TIMER_CSS, DANGER, MUTED, SECONDARY


class CountdownTui(App):
    """A countdown timer app."""

    TITLE = "Timer"
    SUB_TITLE = "Countdown"

    CSS = TIMER_CSS

    BINDINGS = [
        ("q", "quit", "Quit"),
        ("space", "toggle_pause", "Pause/Resume"),
    ]

    time_left = reactive(0.0)

    def __init__(self, seconds: int) -> None:
        super().__init__()
        self.countdown = Countdown(seconds)
        self._finished_announced = False

    def compose(self) -> ComposeResult:
        yield Header(show_clock=True)
        with Container(id="content"):
            with Container(id="display-container"):
                yield Digits("00:00", id="countdown")
                yield Static("Running", id="status")
        yield Footer()

    def on_mount(self) -> None:
        self.time_left = self.countdown.time_left
        self.update_display()
        self._sync_status()
        self.set_interval(0.1, self.tick)

    def tick(self) -> None:
        self.countdown.tick()
        self.time_left = self.countdown.time_left

        if self.countdown.is_finished and not self._finished_announced:
            self._finished_announced = True
            self.notify("Time's up!", severity="error", timeout=10)
            self.bell()
            self.query_one("#status", Static).update("Time's Up!")
            self.query_one("#status", Static).styles.color = DANGER

        self.update_display()
        self._sync_status()

    def update_display(self) -> None:
        time_str = format_time(self.time_left, show_centiseconds=False)
        digits = self.query_one("#countdown", Digits)
        digits.update(time_str)

        # Subtle urgency cue while still respecting the palette.
        if not self.countdown.is_running and not self.countdown.is_finished:
            digits.styles.color = MUTED
        elif self.countdown.is_finished or self.time_left < 10:
            digits.styles.color = DANGER
        else:
            digits.styles.color = SECONDARY

    def action_toggle_pause(self) -> None:
        self.countdown.toggle()
        self._sync_status()

    def _sync_status(self) -> None:
        status_widget = self.query_one("#status", Static)

        if self.countdown.is_finished:
            status_widget.update("Time's Up!")
            status_widget.styles.color = DANGER
            return

        if self.countdown.is_running:
            status_widget.update("Running")
            status_widget.styles.color = SECONDARY
        else:
            status_widget.update("Paused")
            status_widget.styles.color = MUTED

File: src/tui/countdown.ts
import { BoxRenderable, TextAttributes, TextRenderable, type KeyEvent } from "@opentui/core"
import { Countdown } from "../core/termclock.ts"
import { formatTime } from "../core/formatting.ts"
import { createScreen } from "./chrome.ts"
import { renderDigits } from "./digits.ts"
import { PALETTE, mix } from "./theme.ts"

/** Full-screen countdown TUI (faithful port of the original Textual app). */
export async function runCountdownTui(seconds: number): Promise<void> {
  const countdown = new Countdown(seconds)
  let finishedAnnounced = false

  const screen = await createScreen("Countdown", [
    ["q", "Quit"],
    ["space", "Pause/Resume"],
  ])
  const { renderer, content } = screen

  const card = new BoxRenderable(renderer, {
    id: "card",
    flexDirection: "column",
    alignItems: "center",
    border: true,
    borderStyle: "rounded",
    borderColor: mix(PALETTE.gold, 0.2),
    paddingY: 3,
    paddingX: 10,
  })

  const digits = new TextRenderable(renderer, {
    id: "countdown",
    content: renderDigits(formatTime(countdown.timeLeft, { showCentiseconds: false })),
    fg: PALETTE.gold,
  })

  const status = new TextRenderable(renderer, {
    id: "status",
    content: "Running",
    fg: PALETTE.gold,
    attributes: TextAttributes.ITALIC,
    marginTop: 1,
  })

  card.add(digits)
  card.add(status)
  content.add(card)

  // Responsive layout: shrink the card's breathing room as the terminal shrinks.
  screen.onResize((width, height) => {
    card.paddingY = height >= 15 ? 3 : height >= 11 ? 1 : 0
    card.paddingX = width >= 50 ? 10 : width >= 36 ? 4 : 1
  })

  const announceFinished = () => {
    finishedAnnounced = true
    process.stdout.write("\x07") // terminal bell

    // Toast in the top-right corner, like Textual's notify().
    const toast = new BoxRenderable(renderer, {
      id: "toast",
      position: "absolute",
      top: 1,
      right: 2,
      border: true,
      borderStyle: "rounded",
      borderColor: PALETTE.danger,
      backgroundColor: mix(PALETTE.danger, 0.25),
      paddingX: 2,
      zIndex: 10,
    })
    toast.add(
      new TextRenderable(renderer, {
        content: "Time's up!",
        fg: PALETTE.cream,
        attributes: TextAttributes.BOLD,
      }),
    )
    renderer.root.add(toast)
    setTimeout(() => renderer.root.remove(toast), 10_000)
  }

  const sync = () => {
    countdown.tick()
    digits.content = renderDigits(formatTime(countdown.timeLeft, { showCentiseconds: false }))

    const isFinished = countdown.isFinished
    const isPaused = !countdown.isRunning && !isFinished
    const isUrgent = !isPaused && (isFinished || countdown.timeLeft < 10)

    if (isFinished && !finishedAnnounced) announceFinished()

    digits.fg = isUrgent ? PALETTE.danger : isPaused ? mix(PALETTE.gold, 0.7) : PALETTE.gold

    if (isFinished) {
      status.content = "Time's Up!"
      status.fg = PALETTE.danger
    } else if (countdown.isRunning) {
      status.content = "Running"
      status.fg = PALETTE.gold
    } else {
      status.content = "Paused"
      status.fg = mix(PALETTE.gold, 0.7)
    }
  }

  return new Promise((resolve) => {
    let finished = false

    const quit = () => {
      if (finished) return
      finished = true
      clearInterval(timer)
      screen.cleanup()
      renderer.destroy()
      resolve()
    }

    renderer.keyInput.on("keypress", (key: KeyEvent) => {
      if (key.name === "q" || (key.ctrl && key.name === "c")) {
        quit()
      } else if (key.name === "space" || key.sequence === " ") {
        countdown.toggle()
        sync()
      }
    })

    const timer = setInterval(sync, 100)
    sync()
  })
}
